### PR TITLE
Provide a separate IP range for each cluster in the federation.

### PR DESCRIPTION
**Pull request by madhusudancs**
cc @kubernetes/sig-cluster-federation @colhom 

[Analytics]()

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/27812)

### CODE REVIEW

**Comment by quinton-hoole** [comment]
@madhusudancs Would it not be a lot simpler to just pass in a separate IP range/subnet for each cluster in the federation, as an environment variable?

**Comment by madhusudancs** [comment]
@quinton-hoole it is. But then you can't default the values. Users who start the federation are expected to provide the IP address range manually for each cluster.

File: hack/e2e-internal/e2e-up.sh [hunk 1]
```diff
@@ -29,16 +29,40 @@ source "${KUBE_ROOT}/cluster/kube-util.sh"
 
 prepare-e2e
 
+function calc_next_cidr {
```
Comment (colhom) [inline hack/e2e-internal/e2e-up.sh]: @madhusudancs could we do something more readable/ less convoluted here, given our simple use case? 

```sh
MAX_NET_NUMBER=255
CURRENT_NET_NUMBER=1

function calc_next_cidr {
     CURRENT_CIDR="10.180.${CURRENT_NET_NUMBER}.0/16"
     CURRENT_NET_NUMBER=$(( CURRENT_NET_NUMBER+1 ))
}

```
Also, why use a /14 address space? Generally in these cases /16 is as large as you'd even need to go, and I'd even say we could just assume that for simplicity's sake.

A person attempting to decipher this must be **very** comfortable with bash and cidr notation to even begin to understand what you have now. Mad props for the octal foo, but I think it's way overkill and going to confuse way more people than it helps in this situation.

Comment (madhusudancs) [inline hack/e2e-internal/e2e-up.sh]: You can't go to 100K pods with /16 because that only gives us 65K addresses. I don't like dumbing down things just because it looks simple now esp. when we have a better solution that works.

Also, it is not as simple as `CURRENT_NET_NUMBER+=1`. You probably could do that with /16 and

```shell
CURRENT_NET_NUMBER=180

function calc_next_cidr {
    CURRENT_CIDR="10.${CURRENT_NET_NUMBER}.0.0/16"
    ...
}
```

But it gets a little more trickier with /14 addresses. The solution in this PR is the right solution to this problem and it works. We shouldn't try to dumb it down.

Comment (quinton-hoole) [inline hack/e2e-internal/e2e-up.sh]: I would like us to go with one of the two simpler solutions proposed above instead for v1.3.  Like @colhom I'm very uncomfortable with the complex bash scripting in the original version of this PR.  Agree @madhusudancs ?

Comment (madhusudancs) [inline hack/e2e-internal/e2e-up.sh]: @quinton-hoole Disagree, for the reasons above.

Comment (madhusudancs) [inline hack/e2e-internal/e2e-up.sh]: Alright, conceded. Here is the dumbed down version - https://github.com/kubernetes/kubernetes/pull/27892

**Comment by quinton-hoole** [comment]
Keeping open for re-consideration for v1.4.

**Comment by quinton-hoole** [comment]
@madhusudancs Should we close this?

**Comment by madhusudancs** [comment]
@quinton-hoole why? Did we not decide to merge it for 1.4?

**Comment by colhom** [comment]
While this is not necessary for our present suite of e2e tests, it definitely is the right thing to do. Over here we've been toying with the idea of backing our AWS pod networks across federated clusters with a bunch of subnets in the same VPC- which would require separate pod IP ranges for each federation member cluster.

I think having disjoint ip ranges for the subnets will only open possibilities on the varieties of federation architectures we can handle testing.

**Comment by goltermann** [comment]
@madhusudancs this one's well out of date.  Needed for v1.4?

**Comment by madhusudancs** [comment]
@goltermann This would have been a nice-to-have for 1.4.

But I am also working on obsoleting the federation scripts that use this code, so I don't think we will need this. I will close this PR. I will reopen it if necessary.
